### PR TITLE
part1 and part2 of exe1

### PR DESCRIPTION
**Pull request by jenupawan**

### CODE REVIEW

File: message_generoator/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "message_generoator"
```
Comment (shashank676) [inline message_generoator/Cargo.toml]: typo: message_generator

File: message_generoator/proto_file/person.proto [hunk 2]
```diff
@@ -0,0 +1,7 @@
+syntax = "proto3";
+
+message Person{
+  string last_name = 1;
+  string first_name = 2;
+  string dob = 3;
+}
```
Comment (shashank676) [inline message_generoator/proto_file/person.proto]: always have a new line at the end of the file

File: message_generoator/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+extern crate protoc_rust;
+fn main() {
+    protoc_rust::Codegen::new()
+        .out_dir("src/protos/")
+        .inputs(&["proto_file/person.proto"])
+        .include("proto_file")
+        .run()
+        .expect("Running protoc failed.");
```
Comment (shashank676) [inline message_generoator/src/main.rs]: improve error msg

File: proto/protos/.rustc_info.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"rustc_fingerprint":16456912848005604488,"outputs":{"16929990295604814582":["___\n",""],"1164083562126845933":["rustc 1.46.0-nightly (fd4b177aa 2020-06-08)\nbinary: rustc\ncommit-hash: fd4b177aabb9749dfb562c48e47379cea81dc277\ncommit-date: 2020-06-08\nhost: x86_64-unknown-linux-gnu\nrelease: 1.46.0-nightly\nLLVM version: 10.0\n",""],"4476964694761187371":["___\nlib___.rlib\nlib___.so\nlib___.so\nlib___.a\nlib___.so\n/home/pj/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu\ndebug_assertions\nproc_macro\ntarget_arch=\"x86_64\"\ntarget_endian=\"little\"\ntarget_env=\"gnu\"\ntarget_family=\"unix\"\ntarget_feature=\"fxsr\"\ntarget_feature=\"mmx\"\ntarget_feature=\"sse\"\ntarget_feature=\"sse2\"\ntarget_has_atomic=\"16\"\ntarget_has_atomic=\"32\"\ntarget_has_atomic=\"64\"\ntarget_has_atomic=\"8\"\ntarget_has_atomic=\"ptr\"\ntarget_has_atomic_load_store=\"16\"\ntarget_has_atomic_load_store=\"32\"\ntarget_has_atomic_load_store=\"64\"\ntarget_has_atomic_load_store=\"8\"\ntarget_has_atomic_load_store=\"ptr\"\ntarget_os=\"linux\"\ntarget_pointer_width=\"64\"\ntarget_thread_local\ntarget_vendor=\"unknown\"\nunix\n",""]},"successes":{}}
```
Comment (shashank676) [inline proto/protos/.rustc_info.json]: no need to push this file in repo

File: proto/protos/per.proto [hunk 5]
```diff
@@ -0,0 +1,7 @@
+syntax = "proto3";
+
+message Person{
+  string last_name = 1;
+  string first_name = 2;
+  string dob = 3;
+}
```
Comment (shashank676) [inline proto/protos/per.proto]: new line at eof missing

File: proto/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    let matches = App::new("proto")
+        .version("1.0")
+        .about("example of using protofobuff")
+        .author("Pawan Jenu")
+        .arg(
+            Arg::with_name("input")
+                .short("-i")
+                .long("-input_file")
+                .help("path for input file with data")
```
Comment (shashank676) [inline proto/src/main.rs]: "path of the input file" is sufficient

File: proto/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        .arg(
+            Arg::with_name("output")
+                .short("-o")
+                .long("-output_file")
+                .help("path for output file")
+                .takes_value(true),
+        )
+        .get_matches();
+    
+    let input_path = matches.value_of("input").unwrap_or("data/input.txt");
```
Comment (shashank676) [inline proto/src/main.rs]: using unwrap_or is not a good option, if input path is not correct u need to throw error

File: proto/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    let output_path = matches.value_of("output").unwrap_or("data/output.txt");
```
Comment (shashank676) [inline proto/src/main.rs]: same as above

File: proto/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    let input_file = File::open(input_path).expect("file not found");
+    let input_buffer = BufReader::new(input_file);
+    let mut ofile = File::create(output_path).expect("unable to create file");
```
Comment (shashank676) [inline proto/src/main.rs]: use correct naming convention

File: proto/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    for line in input_buffer.lines() {
+        let each_line = line.unwrap();
+        let mut values: Vec<String> = Vec::new();
+        for i in each_line.split(',') {
+            values.push(i.to_string())
+        }
```
Comment (shashank676) [inline proto/src/main.rs]: remember this String uses heap memory whereas &str uses stack memory. stack memory is used for static memory allocation and Heap for dynamic memory allocation, both stored in the computer's RAM. Variables allocated on the stack are stored directly to the memory and access to this memory is very fast, and its allocation is dealt with when the program is compiled.
so it better to use
**let line_info: Vec<&str> = each_line.split(',').collect();**

File: proto/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,54 @@
+extern crate clap;
+extern crate protobuf;
+use clap::{App, Arg};
+use protobuf::*;
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::{BufReader, Write};
+
+mod person;
+use person::Person as person_obj;
```
Comment (shashank676) [inline proto/src/main.rs]: rust is not object oriented, using terminologies like person obj...doesn't make much sense

Comment (shashank676) [reply proto/src/main.rs]: i'll prefer Person as it is

File: proto2/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+version = "0.1.0"
+authors = ["Pawan Jenu <pawanjenu@gmail.com>"]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+protobuf = { version = "2.14.0", features = ["with-bytes"] }
+chrono = "0.4.11"
+time = "0.2.16"
+clap = "2.33.1"
```
Comment (shashank676) [inline proto2/Cargo.toml]: newline at eof missing

File: proto2/data/output.txt [hunk 13]
```diff
@@ -0,0 +1,5 @@
+last_name: "Bartowski" first_name: "Chuck" dob: "1985-10-01" age:0years 5month 1days
+last_name: "Walker" first_name: "Sarah" dob: "1987-07-21" age:0years 5month 1days
+last_name: "Casey" first_name: "John" dob: "1976-04-22" age:0years 5month 1days
+last_name: "Larkin" first_name: "Bryce" dob: "1985-11-12" age:0years 5month 1days
+last_name: "Beckman" first_name: "Diane" dob: "1963-02-28" age:0years 5month 1days
```
Comment (shashank676) [inline proto2/data/output.txt]: age calc looks incorrect

Comment (shashank676) [reply proto2/data/output.txt]: 0 years for different dob how is it possible

File: proto2/src/calc_age.rs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    assert_eq!(calculate_age(test_date), "0years 5month 1days");
+}
+
+#[test]
+#[should_panic(expected = "birth date exeeding current date")]
+fn test2_calculate_age() {
+  let  date = NaiveDate::parse_from_str("2021-01-15", "%Y-%m-%d");
+  let  test_date = Utc.from_local_date(&date.unwrap()).unwrap();
+    calculate_age(test_date);
+}
```
Comment (shashank676) [inline proto2/src/calc_age.rs]: unit test must cover all possible edge cases, otherwise, it's of no use

File: proto2/data/output.txt (hunk 13 above)
Comment (shashank676) [reply proto2/data/output.txt]: this type of cases should be covered in unit testing

Comment (shashank676) [reply proto2/data/output.txt]: how did u verify ur output? @jenupawan

File: proto2/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            .help("path for input file with data")
+            .takes_value(true),
+    )
+    .arg(
+        Arg::with_name("output")
+            .short("-o")
+            .long("-output_file")
+            .help("path for output file")
+            .takes_value(true),
+    )
```
Comment (shashank676) [inline proto2/src/main.rs]: same comments as in prev program

File: proto2/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    .get_matches();
+
+let input_path = matches.value_of("input").unwrap_or("data/input.txt");
+let output_path = matches.value_of("output").unwrap_or("data/output.txt");
```
Comment (shashank676) [inline proto2/src/main.rs]: same here as prev progm

File: proto2/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+let input_file = File::open(input_path).expect("file not found");
+let mut ofile = File::create(output_path).expect("unable to create file");
```
Comment (shashank676) [inline proto2/src/main.rs]: same

File: proto2/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+let mut input_buffer = BufReader::new(input_file);
+    
+let mut stream = protobuf::CodedInputStream::from_buffered_reader(&mut input_buffer);
+
+    while !stream.eof().unwrap_or(false) {
+
```
Comment (shashank676) [inline proto2/src/main.rs:46]: y this new line

File: proto2/src/main.rs [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        let person_from_file:person_obj = stream.read_message().unwrap();
```
Comment (shashank676) [inline proto2/src/main.rs]: did u run cargo fmt?? @jenupawan

Comment (shashank676) [reply proto2/src/main.rs]: ur code is not indented correctly

Comment (shashank676) [reply proto2/src/main.rs]: I mentioned no unwrap() always use expect()

Comment (shashank676) [reply proto2/src/main.rs]: put a cfg condition for checking such violations in the code....at compile time

File: proto2/src/main.rs [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        let  date_only = NaiveDate::parse_from_str(person_from_file.get_dob(), "%Y-%m-%d");
+        let  person_date = Utc.from_local_date(&date_only.unwrap()).unwrap();
```
Comment (shashank676) [inline proto2/src/main.rs]: no unwrap()

Comment (shashank676) [inline proto2/src/main.rs]: check the logic used here

File: proto2/src/calc_age.rs [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+  if dur.num_days() < 0 {
+      panic!("birth date exeeding current date")
+  }
+  let mut current_date =  current.day() as i32;
+  let mut current_month = current.month()as i32;
+  let mut current_year  = current.year()as i32;
+  let  birth_date = person_date.day()as i32;
+  let  birth_month = person_date.month()as i32;
+  let  birth_year = person_date.year()as i32;
+  let month:Vec<i32>= vec![31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
```
Comment (shashank676) [inline proto2/src/calc_age.rs]: plural variable name.....use appropriate var names

File: proto2/src/calc_age.rs [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+  let dur = current.signed_duration_since(person_date);
+  if dur.num_days() < 0 {
+      panic!("birth date exeeding current date")
+  }
+  let mut current_date =  current.day() as i32;
+  let mut current_month = current.month()as i32;
+  let mut current_year  = current.year()as i32;
+  let  birth_date = person_date.day()as i32;
+  let  birth_month = person_date.month()as i32;
+  let  birth_year = person_date.year()as i32;
```
Comment (shashank676) [inline proto2/src/calc_age.rs]: cargo fmt not run

File: proto2/src/calc_age.rs [hunk 23]
```diff
@@ -0,0 +1,45 @@
+use chrono::prelude::*;
+use chrono::{ Date, Utc};
+pub fn calculate_age(person_date: Date<Utc>)-> String{
+  let current = Utc::now().date();
```
Comment (shashank676) [inline proto2/src/calc_age.rs]: use system now, in this case

**Review by shashank676 — CHANGES_REQUESTED** [review]

File: proto/protos/per.proto (hunk 5 above)
Comment (jenupawan) [reply proto/protos/per.proto]: @shashank676 , I think this file is not needed anymore, as I will be generating the .rs file in my message_generator project, so can I delete it ?

File: proto/src/main.rs (hunk 7 above)
Comment (jenupawan) [reply proto/src/main.rs]: @shashank676 , what kind of error message to throw inline 34, as am getting the path from command line?

**Review by shashank676 — APPROVED** [review]

**Comment by shashank676** [comment]
please make logical commits

**Review by shashank676 — CHANGES_REQUESTED** [review]
code is fine now @pawanjenu but 20 commits is not correct. You should be writing logical commits

**Review by shashank676 — APPROVED** [review]
